tool/src/java/mod.rs

```diff
@@ -0,0 +1,1461 @@
+                                MaybeStatic::Static => None,
+                                MaybeStatic::NonStatic(lt) => Some(lt),
+                            })
+                            .map(|lt| {
+                                let lt_edges = lt_env
+                                    .all_longer_lifetimes(&lt)
+                                    .map(|lt| lt_env.fmt_lifetime(lt))
+                                    .map(|lt| format!("{lt}Edges.stream()"))
+                                    .mk_str_iter(", ");
+
+                                format!("Stream.concat({lt_edges}).toList()")
+                            })
+                            .mk_str_iter(", ");
+                        format!("new {ty_name}(arena, {native_val}, {self_edges}, {lt_edges})")
+                            .cown()
+                    }
+                    _ => todo!(),
+                };
+                let ty = self.formatter.fmt_java_type(&field.ty);
+                FieldTpl {
+                    name,
+                    ty,
+                    field_val,
+                }
+            })
+            .collect();
+        let edges = s
+            .lifetimes
+            .lifetimes()
+            .lifetimes()
+            .filter_map(|lt| match lt {
+                MaybeStatic::Static => None,
+                MaybeStatic::NonStatic(lt) => Some(lt),
+            })
+            .map(|lt| s.lifetimes.fmt_lifetime(lt))
+            .collect();
+        let (methods, _) = self.gen_methods(s.name.as_str(), &s.methods);
+        (
+            format!("{type_name}.java").into(),
+            StructTypeTpl {
+                type_name: type_name.into(),
+                lib_name: lib_name.clone(),
+                domain: domain.clone(),
+                edges,
+                fields,
+                methods,
+            }
+            .render()
+            .expect("failed to render struct type"),
+        )
+    }
+
+    fn gen_opaque_def(&self, o: &OpaqueDef) -> (Cow<str>, String) {
+        let Config { domain, lib_name } = &self.tcx_config;
+        let (static_methods, class_methods) = self.gen_methods(o.name.as_str(), &o.methods);
+
+        let edges = o
+            .lifetimes
+            .lifetimes()
+            .lifetimes()
+            .filter_map(|lt| match lt {
+                MaybeStatic::Static => None,
+                MaybeStatic::NonStatic(lt) => Some(lt),
+            })
+            .map(|lt| o.lifetimes.fmt_lifetime(lt))
+            .collect();
+
+        let opaque_tpl = OpaqueTypeTpl {
+            type_name: o.name.to_string().into(),
+            lib_name: lib_name.clone(),
+            domain: domain.clone(),
+            edges,
+            static_methods,
+            class_methods,
+        };
+
+        (
+            format!("{}.java", o.name).into(),
+            opaque_tpl.render().expect("Failed to render opaque type"),
+        )
+    }
+}
+
+trait PostFix: Sized {
```

Comment:
nit: i'd avoid these helpers, just use Ok()/Err() and`.into()` for Cow

Reply:
removed